docs: restructured and added new documentation

sremsen opened the pull request:
Was working directly off main before I decided to create a separate branch and a PR, so now it looks like it won't automatically merge (sorry!).

**General updates**
* Restructured entire documentation (pages are organized in to their source folders that match the nav)
* Added new pages for about melodi, quickstart, etc.
* Added Python and Typescript SDK documentation
* Added API playgrounds for Get Threads, Create Thread (see note below)
* First pass of React component documentation (see note below)
* Updated Open API YAML file with support for GET and POST threads and schemas for threads, messages, etc. and removed all references to logs, samples, and experiments.

**Next up:**
* Create documentation for feedback, user, and project endpoints (update Open API documentation & created detailed documentation)
* Update React component documentation (and get screenshots)

**First pass of React component documentation**
* I took the Readme from the repo and modified it repeatedly based on the current Open API file. It's an AI mismash but figured it would be a decent first pass. I can revisit once I've completed the feedback and user documentation and get some direction from Greg.

**Specific notes for Create Thread**
* I added a request example, but it seems to have overwritten Mintlify’s automatically generated cURL, Python, Typescript, etc. (like on the Create Thread page). I can manually re-add these but it might be messy and I wanted to make sure this was right first
* The API playground UI for this endpoint is extremely confusing because messages is an array of objects. I tried to find a way to easily pass it a stringified json version but because it’s auto-generated by the types set in the open api, this just seems to be the UI generated by an array of objects. Changing the type of the message schema to “json” (vs. array) in the open api provides a nicer UI, but I think is wrong.

gmbrown commented:
Rebased to handle merge conflicts

gmbrown commented on integrating-melodi-analytics/react-component-new.mdx:
```diff
@@ -0,0 +1,173 @@
+
+Here's how your thread will appear in the Melodi dashboard:
+
+![Thread Render](../images/render-thread.png)
```
Just a suggestion, but I think it would be nice if the image here matched the examples above

sremsen replied:
Good point. I'm going to update the examples here to match the standard example we're using everywhere and then add an updated image.

gmbrown commented on integrating-melodi-analytics/react-component-new.mdx:
Ultimately I think we should only have one set of documentation for the react component. Either this file or the readme in https://github.com/Melodi-fyi/melodi-sdk-react - if this is where we want it to live, I think I should updated that readme to just point to these docs rather than have some duplicated stuff over there. I just don't want to have to maintain it in two places and run the risk that they get out of sync

gmbrown replied:
And I know that the readme over there is totally out of date, so I probably should just nuke it sooner rather than later.

gmbrown replied:
https://github.com/Melodi-fyi/melodi-sdk-react/pull/36

sremsen replied:
Let's remove it from the Readme and link back to the docs.melodi.fyi page. I can imagine the Mintlify docs being easier to maintain over time.

gmbrown commented on integrating-melodi-analytics/react-component-new.mdx:
```diff
@@ -0,0 +1,173 @@
+
+### Message Types
+
+Messages in a thread can be one of three types:
```
There are only 2 message types supported, "markdown" and "json"

The role on the message is a separate field and actually can be set to anything you want

gmbrown commented on integrating-melodi-analytics/react-component-new.mdx:
```diff
@@ -0,0 +1,173 @@
+   - Used for structured data or tool outputs
+   - Set `type: 'json'`
+   - Data goes in the `jsonContent` field
+   - Requires `role: 'json_tool'`
```
This isn't actually required - you can set the role to whatever you want

gmbrown commented on integrating-melodi-analytics/react-component-new.mdx:
````diff
@@ -0,0 +1,173 @@
+
+```jsx
+<MelodiFeedbackWidget
+  thread={{
````
This will work and is fine for the first pass, but I think what we ultimately want to get people to do is send us the thread via the typescript sdk and then reference the threadId or externalThreadId in the feedback widget.

The upside of that approach is that we will have all of the threads, including ones that never receive feedback.

The downside is that it involves both the typescript sdk and the react widget and it's ever so slightly more complicated to get up and running.

Happy to have a discussion about it, just thinking out loud here

sremsen replied:
I like it! I added tabbed views so now you can see both

gmbrown reviewed (commented):
I see a lot of docs these days that basically explain concepts and datatypes and have a language switcher that just shows you example code in that language.

I wonder if there is a way to have mintlify do something like that and they we could just have a few sections for create project, create thread, create feedback etc and the code switcher rather than a different section for each sdk. Just a thought - maybe more effort than it's worth at the moment, but it might actually make things cleaner/simpler

Relatedly, there are a couple of concepts that are agnostic to whether you use the api or python or typescript sdks that I think are worth mentioning on their own somewhere.

1) externalThreadId/externalMessageId/externalUserId - these are strings that are passed in and can be used to reference these objects without having to get and store the Melodi ids to retrieve things later. For example, the CreateFeedbackRequest can take an externalThreadId, externalMessageId and externalId on the externalUser object.

2) Possibly user segments and feedback attributes are complicated enough that they deserve a little bit of explanation. (or maybe we just leave them out of these docs for now and we can explain them to people who actually have need/use for them)

gmbrown commented on api-reference/create-thread.mdx:
I'm a little confused about the relationship between this file and the definitions for these requests in openapi.yaml -seems like there is a lot of repeated stuff between the two. Is there a way to get this file to reference the schemas defined in there?

gmbrown replied:
I think we should try autogenerating this page from the openapi file
https://mintlify.com/docs/api-playground/openapi/setup#autogenerate-files

And then adding a few comments and details if necessary.

The reason I think this is the right approach is that it will prove that the openapi file is correct (which is important to us if we think Matt might try to autogenerate a Kotlin client from it) and I think it might handle my comment above

gmbrown replied:
Let me know if that's what you did for this already - I assumed not because the message array here doesn't look like it matches the message schema defined in the openapi file

sremsen replied:
They're pseudo auto-generated, I probably just missed the nuance when I was generating it.

This at the top automatically generates the API playground
`---
api: POST https://app.melodi.fyi/api/external/threads
---`

But then I manually added parameters, etc. 

If you auto-generate it completely, then it doesn't seem like you can edit/customize it (the .mdx file doesn't update with the code or anything). BUT I agree that this is the right approach if we can get the open api file correct. Take a quick look at the Auto Create Thread page and if it looks good with no customization, then let's take that approach for the other endpoints, which I can address in another PR.

sremsen commented on getting-started/quickstart.mdx:
```diff
@@ -0,0 +1,10 @@
+---
+
+You can get started with Melodi in less than 10 minutes. First, you'll need to log conversation data to Melodi, which you can do with the following options:
+* [React component](/integrating-melodi-analytics/react-component)
```
switched all to react-component.mdx

sremsen commented on integrating-melodi-analytics/overview.mdx:
```diff
@@ -0,0 +1,39 @@
+
+Melodi provides a number of ways to integrate with your application. To get started, create a project via the SDKs, the API directly, or the React component Then you can create threads, messages, user feedback, and users.
+
+Create a project
```
huh looks like the headers got lost somewhere along the way - they were headers originally!

sremsen commented:
> I see a lot of docs these days that basically explain concepts and datatypes and have a language switcher that just shows you example code in that language.
> 
> I wonder if there is a way to have mintlify do something like that and they we could just have a few sections for create project, create thread, create feedback etc and the code switcher rather than a different section for each sdk. Just a thought - maybe more effort than it's worth at the moment, but it might actually make things cleaner/simpler
> 
> Relatedly, there are a couple of concepts that are agnostic to whether you use the api or python or typescript sdks that I think are worth mentioning on their own somewhere.
> 
> 1. externalThreadId/externalMessageId/externalUserId - these are strings that are passed in and can be used to reference these objects without having to get and store the Melodi ids to retrieve things later. For example, the CreateFeedbackRequest can take an externalThreadId, externalMessageId and externalId on the externalUser object.
> 2. Possibly user segments and feedback attributes are complicated enough that they deserve a little bit of explanation. (or maybe we just leave them out of these docs for now and we can explain them to people who actually have need/use for them)

This is a great idea! Looks like we can do that with Code Groups. I'll see what I can do.

sremsen commented on integrating-melodi-analytics/integrating-from-slack.mdx:
Good point. I went a little rogue and generated some stuff without a ton of oversight on the "integrate with Slack" page. Let me know if it's helpful or if it's total gibberish and we should take it out.

sremsen commented:
**Updated the PR with these changes:**
* Small language improvements
* Combined Python and Typescript SDKs into one page with different code examples in Code Groups
* Improved formatting on React component page (Added request example to the right and changed to nicer formatted props)
* Consolidated discussion of key concepts and datatypes to the Overview: Concepts and Datatypes page, which removed redundant text from elsewhere and should make it easier to maintain
* Added tabs for widget and widget + SDK to support multiple setup paths on the React Component page
* Changed theme and changed to brand colors
* Added a fully autogenerated Create Thread page - take a look and see if looks good without any customization (seems to)
* Added some AI-generated context around integrating with Slack. If it's helpful, great. But if it's terribly wrong, let's just remove the page completely.

gmbrown commented on openapi.yaml:
```diff
@@ -223,188 +291,56 @@ components:
-          type: string
-        version:
-          type: string
+        - properties:
```
It looks like [there is a way to give these titles](https://mintlify.com/docs/api-playground/openapi/advanced-features#providing-options-with-oneof-and-anyof) so that we can call them something like "Markdown Message" and "Json Message"

I think that would be helpful to make this a little less confusing: 

<img width="635" alt="Screenshot 2024-10-31 at 12 05 52 PM" src="https://github.com/user-attachments/assets/c165140d-be40-4d03-aaec-77009a422152">

gmbrown replied:
Also, once I select the markdown option - ideally the "jsonContent" field wouldn't be there anymore - my guess is the best way to accomplish this is to create two new schemas for JsonMessage and MarkdownMessage and make the message definition just by "oneOf" a reference to those two schemas. Let me know if that doesn't make sense.

Here's what i see now that seems imperfect:

<img width="516" alt="Screenshot 2024-10-31 at 12 07 35 PM" src="https://github.com/user-attachments/assets/158e5f56-4474-46d0-a354-321fff813361">

gmbrown commented on openapi.yaml:
```diff
@@ -33,53 +106,58 @@ paths:
-              $ref: '#/components/schemas/Thread'
+              $ref: '#/components/schemas/CreateThreadRequest'
       responses:
         '201':
```
Right now this is actually a 200, but I can update on the backend to a 201 since that is technically more correct

gmbrown reviewed (commented):
Thanks for all your work on this. A few comments about the autogenerated POST Thread page, mostly small things - the good news is that I was able to get it to actually successfully create a thread pretty easily!

sremsen commented on openapi.yaml:
```diff
@@ -484,3 +431,61 @@ components:
+        - messages
+      anyOf:
+        - required: [projectId]
+        - required: [projectName]
```
Apparently it's a limitation of the UI <> OpenAI schema. I updated the OpenAI schema to be longer but support the correct UI:

`CreateThreadRequest:
  oneOf:
    - type: object
      required: [messages, projectId]
      properties:
        projectId:
          type: number
        projectName:
          type: string
          nullable: true
        // ... other properties ...
    - type: object
      required: [messages, projectName]
      properties:
        projectId:
          type: number
          nullable: true
        projectName:
          type: string
        // ... other properties ...`

sremsen commented:
**Updates:**
* Addressed Greg's comments (added titles in the schema, and separated messages into two schemas, updated references to anyOf)
* For Get Threads, I updated Ids, externalUserIds, intentIds, and userSegmentIds to use oneOf pattern to support an integer and an array of integers
* Hid the "slack integration" from the nav (we can come back it when we're ready to implement it ourselves)
* Deleted the original Create Thread and Get Thread pages and replaced with fully auto-generated pages
